src/Ng/ElfieFromCatalogCollector.cs

```diff
@@ -339,35 +358,463 @@ string CreateIdxFile(string packageId, string packageVersion, string tempDirecto
+
+                        // Wait for a few seconds before retrying.
+                        int delay = Catalog2ElfieOptions.GetRetryDelay(retry);
+                        Thread.Sleep(delay * 1000);
+
+                        Trace.TraceWarning($"Retrying getting latest stable version.");
+                    }
+                    else
+                    {
+                        // We've tried a few times and still failed. 
+                        // Rethrow the exception so we can track the failure.
+                        throw;
+                    }
+                }
+            }
+
+            if (latestStableVersion == null)
+            {
+                Trace.TraceWarning($"No latest stable version {packageId}.");
+            }
+            else
+            {
+                Trace.TraceInformation($"Latest stable version {packageId} {latestStableVersion.CatalogEntry.PackageVersion}.");
+            }
+
+            return latestStableVersion;
+        }
+
+        /// <summary>
+        /// Filters the package list to only the latest stable version of the packages
+        /// </summary>
+        /// <param name="packageDownloadCounts"></param>
+        /// <param name="downloadCountThreshold"></param>
+        /// <returns></returns>
+        IEnumerable<RegistrationIndexPackage> FilterPackagesToInclude(Dictionary<string, long> packageDownloadCounts, long downloadCountThreshold)
+        {
+            // In general, here's how the filtering works.
+            //   1. Calcuate the total download count for all the NuGet packages. (this is done before calling this method.)
+            //   2. Calculate how many downloads we want to include in the ardb. (this is done before calling this method.)
+            //   3. Exclude any packages which do not have a latest stable version.
+            //   4. Sort the packages based on download count.
+            //   5. Include the most popular packages until the target download count is reached.
+            //   6. Sort the included packages, first by log2(downloadcount), then by package name.
+            //      This sort order ensures the most popular packages are recommended first, while
+            //      minimizing the day-to-day differences.
+
+
+            // The runninng count of the downloads included in the ardb.
+            long includedDownloadCount = 0;
+
+            // A flag to signal that the download threshold was reached.
+            bool thresholdReached = false;
+            List<Tuple<RegistrationIndexPackage, int>> includedPackagesWithCounts = new List<Tuple<RegistrationIndexPackage, int>>();
+
+            // We need to determine the latest stable version for each package. But this is a fairly expensive operation since
+            // it makes a network call. We'll process the packages in chuncks so we can get the latest stable version of
+            // the packages using multiple threads.
+            int batchSize = 500;
```

Comment:
I don't think the batch size will change often, but putting it in the config file is a good idea. I'll make the change.